pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -446,155 +413,86 @@ func (m *ManagerImpl) readCheckpoint() error {
 	return nil
 }
 
-// updateAllocatedDevices gets a list of active pods and then frees any Devices that are bound to
-// terminated pods. Returns error on failure.
-func (m *ManagerImpl) updateAllocatedDevices(activePods []*v1.Pod) {
-	if !m.sourcesReady.AllReady() {
-		return
-	}
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-	activePodUids := sets.NewString()
-	for _, pod := range activePods {
-		activePodUids.Insert(string(pod.UID))
-	}
-	allocatedPodUids := m.podDevices.pods()
-	podsToBeRemoved := allocatedPodUids.Difference(activePodUids)
-	if len(podsToBeRemoved) <= 0 {
-		return
-	}
-	glog.V(3).Infof("pods to be removed: %v", podsToBeRemoved.List())
-	m.podDevices.delete(podsToBeRemoved.List())
-	// Regenerated allocatedDevices after we update pod allocation information.
-	m.allocatedDevices = m.podDevices.devices()
-}
+// Allocate allocates a set of device resources for the Pod given.
+func (m *ManagerImpl) Allocate(node *schedulercache.NodeInfo, attrs *lifecycle.PodAdmitAttributes) error {
 
-// Returns list of device Ids we need to allocate with Allocate rpc call.
-// Returns empty list in case we don't need to issue the Allocate rpc call.
-func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, required int) (sets.String, error) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-	needed := required
-	// Gets list of devices that have already been allocated.
-	// This can happen if a container restarts for example.
-	devices := m.podDevices.containerDevices(podUID, contName, resource)
-	if devices != nil {
-		glog.V(3).Infof("Found pre-allocated devices for resource %s container %q in Pod %q: %v", resource, contName, podUID, devices.List())
-		needed = needed - devices.Len()
-		// A pod's resource is not expected to change once admitted by the API server,
-		// so just fail loudly here. We can revisit this part if this no longer holds.
-		if needed != 0 {
-			return nil, fmt.Errorf("pod %v container %v changed request for resource %v from %v to %v", podUID, contName, resource, devices.Len(), required)
+	// Purge allocations by active Pods
+	if m.sourcesReady.AllReady() {
+		activePods := sets.NewString()
+		for _, p := range m.activePods() {
+			activePods.Insert(string(p.UID))
 		}
+		m.allocation.purge(activePods)
```

Comment:
Looks like the return value is not used here.
Do we need to create a map or is it intended for testing purposes ?

Reply:
Right, the return value becomes useless now. The return value was used to initialize the allocated_devices variable previously. So ... we can safely remove it now.